src/main/java/com/amazonaws/encryptionsdk/internal/Utils.java

```diff
@@ -279,4 +280,30 @@ public static ByteBuffer limit(final ByteBuffer buff, final int newLimit) {
     public static String encodeBase64String(final byte[] data) {
         return Base64.toBase64String(data);
     }
+
+    /**
+     * Removes the leading zero sign byte from the byte array representation of a BigInteger (if present)
+     * and left pads with zeroes to produce a byte array of the given length.
+     * @param bigInteger The BigInteger to convert to a byte array
+     * @param length The length of the byte array
+     * @return The byte array
+     */
+    public static byte[] bigIntegerToByteArray(final BigInteger bigInteger, final int length) {
+        final byte[] result = new byte[length];
+        byte[] rawBytes = bigInteger.toByteArray();
+
+        //Remove sign byte if one is present
+        if(rawBytes[0] == 0) {
+            rawBytes = Arrays.copyOfRange(rawBytes, 1, rawBytes.length);
+        }
+
+        if(length < rawBytes.length) {
+            throw new IllegalArgumentException("Length must be as least as long as the BigInteger byte array " +
+                    "without the sign byte");
+        }
+
+        System.arraycopy(rawBytes, 0, result, length - rawBytes.length, rawBytes.length);
+
+        return result;
+    }
```

Comment:
Let's simplify this (and reduce unneeded allocations).

```
final byte[] rawBytes = bigInteger.toByteArray();
// If we're already the correct length, just return it.
if (rawBytes.length ==length) {
    return rawBytes;
}
// If we're exactly one byte too large, but we have a leading zero byte, remove it and return.
if (rawBytes.length == length + 1 && rawBytes[0] == 0) {
    return Arrays.copyOfRange(rawBytes, 1, rawBytes.length);
}

if (rawBytes.length > length) {
    throw new IllegalArgumentException("Length must be as least as long as the BigInteger byte array " +
      "without the sign byte");
}
final byte[] paddedResult = new byte[length];
System.arraycopy(rawBytes, 0, paddedResult , length - rawBytes.length, rawBytes.length);
return paddedResult;
}
```

Reply:
Done